### PR TITLE
PR: Enable scrolling past the end of the document

### LINKED ISSUE #3345: Feature suggestion:  scroll over the end of file in the editor
Currently, if you have 100 lines of code, the scroller won't go over line 100, which means if you want to add more code, you are always typing at the bottom of the editor window (or you fanatically press enters until enough carriage returns are added to the end of file...). 

Please allow scrolling over to the eof, so that user can always work in the middle of screen.

P.S. the visual studio and VS Code has this feature.

### PR DESCRIPTION
**Pull request by Prikers**
Fixes #3345

Hi team, 
This is an attempt to solve #3345 and #2063: I added an option (default to `False`) to enable the user to scroll past the end of the document in order to have the last line of code on top of the screen (without having to added unnecessary blank lines). 
I don't know if this is an option you are willing to add to Spyder. I find it very useful but that is not my call :-)
The option is already available in QtPlainTextEdit (`setCenterOnScroll(True)`) so it did not add a lot of complexity to the editor code. The `scrollflagarea` does not need to be modified.

scrollpastend

For some reason however I can't find a way to update the editor immediatly after adding the option. The tab that has focus is not updated while the others are updated. As soon as I add a new line (or remove one) it gets updated. Could you point me to what I am missing? 
- [x] Update editor after adding the option

### CODE REVIEW

**Comment by Prikers** [comment]
I accidentally commited the removal of all trailling white spaces (automatically done with spyder). Should I revert that?

**Comment by ccordoba12** [comment]
Yep, please revert.

File: spyder/widgets/sourcecode/codeeditor.py [hunk 1]
```diff
@@ -1201,6 +1207,15 @@ def set_blanks_enabled(self, state):
...
         self.rehighlight()
 
+    def set_scrollpastend_enabled(self, state):
+        """
+        Allow user to scroll past the end of the document to have the last
+        line on top of the screen
+        """
+        self.scrollpastend_enabled = state
+        self.setCenterOnScroll(state)
+        self.update()
```
Comment (jnsebgosselin) [inline spyder/widgets/sourcecode/codeeditor.py]: replace `self.update()` by `self.setDocument(self.document())` and it should solve your issue about the option not being applied at runtime.

Comment (Prikers) [reply spyder/widgets/sourcecode/codeeditor.py]: Works like a charm! Thanks @jnsebgosselin!

**Comment by ccordoba12** [comment]
@Prikers, please create a new branch and PR for your last work (the Pandas Index one) to not mix things with this PR. Thanks!

**Comment by Prikers** [comment]
So sorry! I forgot that I had already pushed this first PR...
Could you help me on that one? I was thinking doing a git reset to go back to 7cefc10 and then `git push origin --force` (as I cannot simply push to origin). Is that a bad practice? 
Again all appologies :confused:

**Comment by rlaverde** [comment]
>Could you help me on that one? I was thinking doing a git reset to go back to 7cefc10 and then git push origin --force (as I cannot simply push to origin). Is that a bad practice?

Yes, It's somehow a bad practice, but It's the only way to remove those commits, remember to be very careful when using `--force`

And remember to create new branches for each PR.

**Comment by Prikers** [comment]
ok. Sorry again about that. 
It is now clean and (I think) ready.

**Comment by ccordoba12** [comment]
@Prikers, there are conflicts now. Please merge with master to solve it.

**Comment by Prikers** [comment]
@ccordoba12, @dalthviz, @jnsebgosselin, @rlaverde thanks for considering, reviewing and helphing with this PR! I have merge upstream/master, there are no conflicts anymore. 
As for the pep8 issue, I am not sure how I should handle it... If I add some indentation to [Line999](https://github.com/spyder-ide/spyder/blob/7cefc101f61197a9c39887458181ce7c51d8d088/spyder/plugins/editor.py#L999) for visual indent, line 1000 will not be aligned anymore. And because of the 79 characters limit I can not align line 1000. I tried to stick to the style of the previous lines ([Line991](https://github.com/spyder-ide/spyder/blob/7cefc101f61197a9c39887458181ce7c51d8d088/spyder/plugins/editor.py#L991) for example). @dalthviz what would you suggest?

**Comment by dalthviz** [comment]
@Prikers in that case maybe try writing the `create_action` like this:

```python
        self.scrollpastend_action = create_action(
                                        self,
                                        _("Scroll past the end"),
                                        toggled=self.toggle_scroll_past_end)
```
Note that in the first line (when openning the create_action) you don't have any arguments. I think with this the code style issue will be fixed :). Here a [link to pep8 for reference](http://legacy.python.org/dev/peps/pep-0008/#indentation)

**Comment by ccordoba12** [comment]
Thanks @Prikers for your contribution!

**Comment by goanpeca** [comment]
@Prikers I invited you to the [spyder/plubic](https://gitter.im/spyder-ide/public) chat on gitter. Could you log in when you have some time :-) ?

**Comment by Prikers** [comment]
I have just joined 😄 . 
Thanks @goanpeca !

**Pull request merged**